vehicles/vehiclelighting.cpp

```diff
@@ -0,0 +1,100 @@
+/*
+ *     Copyright 2023 Rickard Häll   rickard.hall@ri.se
+ *     Published under GPLv3: https://www.gnu.org/licenses/gpl-3.0.html
+ *
+ *     Code adapted from example:
+ *     https://www.ics.com/blog/gpio-programming-exploring-libgpiod-library
+ */
+
+#include "vehiclelighting.h"
+
+VehicleLighting::VehicleLighting(QObject *parent)
+    : QObject{parent}
+{
+    // Open GPIO chip
+    if (!(mChip = gpiod_chip_open_by_name(mChipName)))
+        throw QString("Could not open gpio chip with name: ")+=QString::fromLocal8Bit(mChipName);
+    else {
+        // Open GPIO lines
+        mBackupLight = gpiod_chip_get_line(mChip, 22);
+        mBrakeLight = gpiod_chip_get_line(mChip, 23);
+        mRightTurnSignal = gpiod_chip_get_line(mChip, 24);
+        mLeftTurnSignal = gpiod_chip_get_line(mChip, 25);
+
+        // Open LED lines for output
+        gpiod_line_request_output(mBackupLight, "lights", 0);
+        gpiod_line_request_output(mBrakeLight, "lights", 0);
+        gpiod_line_request_output(mRightTurnSignal, "lights", 0);
+        gpiod_line_request_output(mLeftTurnSignal, "lights", 0);
+    }
+}
+
+VehicleLighting::~VehicleLighting()
+{
+    // Release lines and chip
+    gpiod_line_release(mBackupLight);
+    gpiod_line_release(mBrakeLight);
+    gpiod_line_release(mRightTurnSignal);
+    gpiod_line_release(mLeftTurnSignal);
+    gpiod_chip_close(mChip);
+}
+
+void VehicleLighting::turnSignal(double steering)
+{
+    if (steering < -0.5)
+        turnSignalRelay(true);
+    else if (steering > 0.5)
+        turnSignalRelay(false);
+    else
+        stopSignaling();
+}
+
+void VehicleLighting::rearLights(double speed)
+{
+    if (speed < -0.01) {
+        backupLight(true);
+        brakeLight(false);
+    } else if (speed > 0.01) {
+        backupLight(false);
+        brakeLight(false);
+    } else {
+        backupLight(false);
+        brakeLight(true);
+    }
+}
+
+void VehicleLighting::stopSignaling()
+{
+    gpiod_line_set_value(mLeftTurnSignal, 0);
+    gpiod_line_set_value(mRightTurnSignal, 0);
+}
+
+void VehicleLighting::turnSignalRelay(bool left)
+{
+    static int x, counter = 1;
+    if (counter == 10) {
```

Comment:
It looks like how fast the turn signal will blink depends on the assumption that turnSignal is called periodically with the right period. Which in turn creates the assumption that CarMovementController::setDesiredSteering is called with the right period. I do not think that it is an assumption that we can or want make. If turnSignal needs to be called with a certain period to look nice, VehicleLighting should take care of it internally, e.g., with a timer.

Reply:
It now has an internal timer